### ISSUE TITLE
add a JSON Schema for the JSON output

### ISSUE THREAD

**Issue by monperrus** [SpoonLabs/sorald#246]
@khaes-kth @fermadeiral will meet to create first alpha version

**Comment by monperrus** [comment]
We would need two JSON schemas one for miner and one for repair

**Comment by slarse** [comment]
This is related to #225. To be able to do targeted repair, I need the miner schema to contain at least the following for each rule violation:

* The key of the rule.
* Path to the file
    - For portability, should probably be relative to `--originalFilesPath`, i.e. the project root.
* The precise position of the violation (start line, end line, start column, end column)

**Comment by fermadeiral** [comment]
> The precise position of the violation (start line, end line, start column, end column)

If I recall correctly, Sonar Java returns only the file name and the line number of the violation. Was that changed in newer versions?

**Comment by slarse** [comment]
> If I recall correctly, Sonar Java returns only the file name and the line number of the violation. Was that changed in newer versions?

The exact location is semi-hidden in a nested data structure in the `AnalyzerMessage`, see https://github.com/SpoonLabs/sorald/blob/7e01bb9816f7b750464a821716a818ac148c761f/src/main/java/sorald/sonar/ScannedViolation.java#L13-L31

**Comment by khaes-kth** [comment]
@fermadeiral and I prepared a first version of the json schema. The first link below is the schema/example for mining output and the second one is for repair output (including small changes that I made after we made a first draft with Fernanda). They are very close to what @slarse had proposed before.

[https://www.jsonschemavalidator.net/s/Glo83TcP](https://www.jsonschemavalidator.net/s/Glo83TcP)
[https://www.jsonschemavalidator.net/s/afS0RbsJ](https://www.jsonschemavalidator.net/s/afS0RbsJ)

To have an idea with a quick look, have a look at the examples (right hand side).

**Comment by slarse** [comment]
@khaes-kth I've got a question on the mining schema, what's "miningTime" here?

```
            "rules": [
                {
                    "ruleNumber": "2111",
                    "ruleName": "BigDecimalDoubleConstructor",
                    "nbWarningsFound": 1,
                    "warnings": {
                        "location": {
                            "filePath": "relative/path/to/file",
                            "beginningLineNumber": 10,
                            "endLineNumber": 12,
                            "beginningColumnNumber": 3,
                            "endColumnNumber": 1
                        }
                    },
                    "miningTime": 1323
                }
            ],
```
This looks like it's supposed to be the time for a specific warning of a specific rule. We don't have access to any timing information but the total amount of time to mine all considered rules.

**Comment by khaes-kth** [comment]
@slarse it is supposed to show the total time spent for mining all violations of this specific rule (here "2111"). If we do not have access to this time, we can just keep the total time.

**Comment by slarse** [comment]
@khaes-kth Right, okay. But no, we don't have that time, as we pass all checks at once to the miner. Previously we passed one check at a time, but that was very, very slow as Sonar had to re-parse the entire project each time (for large projects it was orders of magnitude slower). We can only get the total amount of time spent mining with Sonar right now.

**Comment by khaes-kth** [comment]
@slarse Good, so let's just keep the total time.
